custom_components/localtuya/fan.py

```diff
@@ -74,38 +92,60 @@ async def async_turn_off(self, **kwargs) -> None:
 
     async def async_set_speed(self, speed: str) -> None:
         """Set the speed of the fan."""
+        self._speed = speed
+
         if speed == SPEED_OFF:
             await self._device.set_dps(False, "1")
-        elif speed == SPEED_LOW:
-            await self._device.set_dps("1", "2")
-        elif speed == SPEED_MEDIUM:
-            await self._device.set_dps("2", "2")
-        elif speed == SPEED_HIGH:
-            await self._device.set_dps("3", "2")
+        else:
+            await self._device.set_dps(
+                self._config.get(speed), self._config.get(CONF_FAN_SPEED_CONTROL)
+            )
+
         self.schedule_update_ha_state()
 
     async def async_oscillate(self, oscillating: bool) -> None:
         """Set oscillation."""
-        await self._device.set_dps(oscillating, "8")
+        await self._device.set_dps(
+            oscillating, self._config.get(CONF_FAN_OSCILLATING_CONTROL)
+        )
         self.schedule_update_ha_state()
 
     @property
     def supported_features(self) -> int:
         """Flag supported features."""
-        return SUPPORT_SET_SPEED | SUPPORT_OSCILLATE
+        if self.has_config(CONF_FAN_OSCILLATING_CONTROL):
+            return SUPPORT_SET_SPEED | SUPPORT_OSCILLATE
+        else:
+            return SUPPORT_SET_SPEED
 
     def status_updated(self):
         """Get state of Tuya fan."""
-        self._is_on = self._status["dps"]["1"]
-        if not self._status["dps"]["1"]:
-            self._speed = SPEED_OFF
-        elif self._status["dps"]["2"] == "1":
-            self._speed = SPEED_LOW
-        elif self._status["dps"]["2"] == "2":
-            self._speed = SPEED_MEDIUM
-        elif self._status["dps"]["2"] == "3":
-            self._speed = SPEED_HIGH
-        self._oscillating = self._status["dps"]["8"]
+        mappings = {
+            self._config.get(SPEED_LOW): SPEED_LOW,
+            self._config.get(SPEED_MEDIUM): SPEED_MEDIUM,
+            self._config.get(SPEED_HIGH): SPEED_HIGH,
+        }
+
+        self._is_on = self.dps(self._dps_id)
+
+        if self.has_config(CONF_FAN_SPEED_CONTROL) and self.dps_conf(
+            CONF_FAN_SPEED_CONTROL
+        ):
+            try:
+                self._speed = mappings[self.dps_conf(CONF_FAN_SPEED_CONTROL)]
```

Comment:
You can avoid the try-catch like this:

```python
self._speed = mappings.get(self.dps_conf(CONF_FAN_SPEED_CONTROLLER))
if self._speed is None:
    _LOGGER.warning(...)
```

Reply:
I keep forgetting about the '.get()' stuff..  I must be too old school I guess.

Reply:
It's nice in some cases, but it's important to think about the potential consequences as well. If a value is expected to be there, e.g. if an option is marked as `vol.Required`, you should still access via index instead of get. If you get an exception you've encountered a bug and you should not hide that. Fail fast and hard. It's always a risk potentially introducing an unwanted `None` value. Just a parenthesis.